### PR TITLE
#04 - Google auth research and implement

### LINKED ISSUE #4: Research & Implement User Authentication
- [x] Read the article for NextAuth & google (https://www.telerik.com/blogs/how-to-implement-google-authentication-nextjs-app-using-nextauth, https://next-auth.js.org/providers/google). Feel free to explore other sources if you want to.
- [x] Implement Google Sign in
- Register with Google Cloud Console, store keys in .env
- Set up NextAuth API Route Providers in api/auth/[...nextauth].js
- Wrap root HTML tag with NextAuth Session Provider (layout.tsx)
- Change the base page.tsx to test signing in (signIn session from next-auth/react using google provider)
- Change the base page.tsx to check session (UseSession hook) for login

### PR DESCRIPTION
**Pull request by kaseyliu**
## Developer: Brandon Eng, Kasey Liu

Closes #04

### Pull Request Summary

Implement Google Sign in:
- Register with Google Cloud Console, store keys (GOOGLE_ID and GOOGLE_SECRET) in .env
- Set up NextAuth API Route Providers in api/auth/[...nextauth]/route.ts
- Wrap root HTML tag with NextAuth Session Provider (layout.tsx)
- Change the base page.tsx to test signing in (signIn session from next-auth/react using google provider)
- Change the base page.tsx to check session (UseSession hook) for login

### Modifications
- changed file app/api/[...nextauth].js into directory, and added route.tsx 
- added Login.tsx: uses Google auth's signIn() and signOut() functions and display sign in/sign out buttons
- added userCard.tsx: to show currently logged in user
- added components/NextAuthProvider.tsx: helper file to use client functions

### Testing Considerations

- Verify that user can sign in with their Google account 
- Verify that user card correctly displays the signed in user's information
- Verify that sign out button appears after signing in

### Pull Request Checklist

- [x] Code is neat, readable, and works
- [x] Comments are appropriate
- [x] The commit messages follows our [guidelines](https://h4i.notion.site/Conventional-Commits-593452ad1179489399ad3bd696ef772a)
- [x] The developer name is specified
- [x] The summary is completed
- [x] Assign reviewers

### Screenshots/Screencast
Screenshot 2024-01-28 at 6 25 22 PM

After clicking "sign in with google":
Screenshot 2024-01-28 at 6 25 54 PM

Screenshot 2024-01-28 at 6 26 24 PM

Displays currently logged in user:
Screenshot 2024-01-28 at 6 26 40 PM

### CODE REVIEW

File: src/app/Login.tsx [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+        return (
+            <>
+                <button onClick={() => signOut()} type="button" className="btn btn-primary">Sign Out of Google</button>
+                <UserCard user={session?.user}/>
+            </>
+        )
+    } else { //if not logged in, sign in
+        return (
+            <>
+                <button onClick={() => signIn()} type="button" className="btn btn-primary">Sign In with Google</button>
```
Comment (andrewtchan) [inline src/app/Login.tsx:18]: Please add the 'google' parameter to the `signIn()` function so we skip the intermediate page and go straight to the Google sign in!

File: package.json
Comment (andrewtchan) [inline package.json:1]: Not sure if npm install generated these changes to package.json and package-lock.json (if so it's fine for this time as it was just a small update to our next version), but in general try to not commit changes to either of these files unless strictly necessary, i.e. for a library we need or smth. If we find a change is required, please separate it into it's own commit in the future.

File: src/app/api/auth/[...nextauth]/route.tsx [hunk 2]
```diff
@@ -0,0 +1,25 @@
+import NextAuth from 'next-auth';
+import GoogleProvider from "next-auth/providers/google";
+import { SessionOptions } from 'next-auth';
+
+
+const googleClientId = process.env.GOOGLE_ID;
+const googleClientSecret = process.env.GOOGLE_SECRET;
+
+if (!googleClientId || !googleClientSecret) {
+    console.log("sad");
```
Comment (andrewtchan) [inline src/app/api/auth/[...nextauth]/route.tsx:10]: hate to be no fun.. but could you pls change this to smth like "missing google auth secrets in env"

File: src/app/layout.tsx [hunk 3]
```diff
@@ -11,10 +12,14 @@ export default function RootLayout({
...
-    <html lang="en">
-      <body>{children}</body>
-    </html>
+    <NextAuthProvider>
+
+      <html lang="en">
+        <body>{children}</body>
+      </html>
+
+    </NextAuthProvider>
```
Comment (andrewtchan) [inline src/app/layout.tsx:23]: remove extra newlines

File: src/app/page.tsx [hunk 4]
```diff
@@ -16,12 +20,17 @@ export default function Home() {
...
+  <main>
+    <Navbar />
+    <h1>Home</h1>
+    <button onClick={handleApiCall}>Test Database Connection</button>
+    <p>API Response: {apiResponse}</p>
+    <div style={{width: "500px", margin: "0 auto", paddingTop: "30px"}}>
+      <h3>Login Website</h3>
+      <Login />
+    </div>
+  </main>
```
Comment (andrewtchan) [inline src/app/page.tsx:33]: indent this block so it's not on the same line as the `Home()` export line

**Review by andrewtchan — CHANGES_REQUESTED** [review]
Great work Kasey and Brandon! I'd personally consider this to be a challenging issue and y'all killed it. Please review my comments as I requested a couple very minor changes before merging this. Also, thank you for the detail in your PR and screenshots -- really helpful.

**Comment by andrewtchan** [comment]
Actually, I went ahead and did the formatting fixes for you guys. I'll talk to Adrian about adding an auto-formatter to our repo but for now pls install a code formatter in your IDE like Prettier (for VSCode).

File: src/app/Login.tsx (hunk 1 above)
Comment (hnadrian) [reply src/app/Login.tsx:24]: @kaseyliu @Bdon101 Make sure you handle this small change if you have time and I'll approve your pull request!

**Review by hnadrian — APPROVED** [review]
All looking good!

**Pull request merged**